### ISSUE TITLE
Memory optimization if reader is an `io.ReaderAt`

### ISSUE THREAD

**Issue by joelanford** [nlepage/go-tarfs#15]
I've been a user of this library, and I find it really useful (thanks!).

However, I've noticed that the files within the tar are fully read into memory, even if the underlying reader supports `io.ReaderAt`. If the underlying reader _does_ support `io.ReaderAt`, it seems like it would be a significant memory optimization to use `io.SectionReader` instead of `bytes.Buffer`, so that `fs.FS` operations could read file contents directly out of the reader.

Is this something that would be possible to explore in this library?

**Comment by joelanford** [comment]
Oh nice! I see you've been exploring this already in https://github.com/nlepage/go-tarfs/pull/14.

Is there something (other than time/effort) that precludes getting that in?

**Comment by nlepage** [comment]
Hi @joelanford 
Yes reading all files in memory is **the problem** about this lib!

It all comes from the [`archive/tar` package's `Reader`](https://pkg.go.dev/archive/tar#Reader), which doesn't support `Seek`, even if the underlying `io.Reader` does support it...

Two solutions have been explored to solve this:
 - *hacking* the standard library's `archive/tar` package to support `Seek` (you can read more about this at #13)
 - the solution you saw at #14

Neither of these is ideal, I'm not happy with #14 because it is cumbersome and benchmarking it gave bad results.

That's why I prefered letting this lib as it is for now, surely not suited for reading large files, but a good and simple solution for reading reasonably sized files.

Could you tell me more about how you are using this lib?
Are you reading large tar files? Are you reading many small tar files? Are you keeping them open in memory on long term (eg. using it to respond to HTTP requests)? Is memory usage really a problem?
Thanks in advance!

**Comment by joelanford** [comment]
> Could you tell me more about how you are using this lib?

Essentially, I find the `fs.FS` abstraction much nicer for passing around "stuff that represents a filesystem", which for my cases typically boils down to tar files (`tarfs.New()`), actual directories trees (`os.DirFS()`), and test filesystems (`fstest.MapFS`). Once the `fs.FS` abstraction was added to the standard library, I started moving function signatures towards accepting this abstraction instead of actual directory paths or actual tar files.

In general, I use this library somewhat ephemerally (i.e. a function that accepts an `fs.FS` and extracts something interesting from it, or converts it to some other representation). Off the top of my head, I can't think of any cases where I hold onto a tarfs indefinitely or for long periods of time.

> Are you reading large tar files? Are you reading many small tar files? Are you keeping them open in memory on long term (eg. using it to respond to HTTP requests)? Is memory usage really a problem?

Size wise, I'm currently dealing with <1GB tar files. So not huge, and things generally fit in memory on most machines, but I'm trying to avoid large spikes in memory. One use case is to be use this library in a Kubernetes context where I can set pod resource limits for memory just above expected steady state of the process.

---

I started scratching the surface of what it would take to do this, and I ended up with something that looked roughly like this pseudocode:
```go
type entry struct {
	dataStartLocation int64 // get this using a custom reader that keeps track of bytes read by the tar.Reader
	tar.Header
}

func (tfs *tarfs) Open(name) (fs.File, error) {
	entry, err := tfs.entries[name]
	if err != nil {
		return nil, err
	}

    // if the passed in reader is not a readerAt, read the entire reader into a byte buffer
    // and assign tfs.readerAt as the byte buffer. from then on, everything else can assume
    // we have a readerAt.
	reader := io.NewSectionReader(tfs.readerAt, entry.dataStartLocation, entry.Size)
	return  &file{r: reader, ... /* and other stuff as needed */}, nil
}
```

I say this with a grain of salt though. I haven't actually played this idea out far enough myself to see what problems would crop up.
